# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/gc/FinalizationObservers.cpp
# Commit: d30323c51a42
# Full Hash: d30323c51a42996c0cbab0597f15b7abf57b720a
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2022-03-04 15:59:45
# Description:
#   Bug 1757573 - Use the correct zone when looking up nuked WeakRef CCWs r=sfink
#   
#   WeakRefs are exposed to the user so anyone can create CCWs to WeakRefs. Don't
#   assume that a nuked WeakRef CCW is one that we created but get it's target to
#   find which zone to look it up in.
# ==============================================================================

diff -r 4bcb69527b98 -r d30323c51a42 js/src/gc/FinalizationObservers.cpp
--- a/js/src/gc/FinalizationObservers.cpp	Fri Mar 04 04:44:07 2022 -0500
+++ b/js/src/gc/FinalizationObservers.cpp	Fri Mar 04 10:08:31 2022 +0000
@@ -352,7 +352,15 @@
 }
 
 void GCRuntime::nukeWeakRefWrapper(JSObject* wrapper, WeakRefObject* weakRef) {
-  FinalizationObservers* observers = wrapper->zone()->finalizationObservers();
+  // WeakRef wrappers can exist independently of the ones we create for the
+  // weakRefMap so don't assume |wrapper| is in the same zone as the WeakRef
+  // target.
+  JSObject* target = weakRef->target();
+  if (!target) {
+    return;
+  }
+
+  FinalizationObservers* observers = target->zone()->finalizationObservers();
   if (observers) {
     observers->unregisterWeakRefWrapper(wrapper, weakRef);
   }